x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java

```diff
@@ -109,137 +110,184 @@ protected void masterOperation(
         RollupAction.Request request,
         ClusterState state,
         ActionListener<AcknowledgedResponse> listener
-    ) throws IOException {
-        String originalIndexName = request.getSourceIndex();
-
-        final String rollupIndexName;
-        if (request.getRollupIndex() == null) {
-            rollupIndexName = "rollup-" + originalIndexName + "-" + UUIDs.randomBase64UUID(Randomness.get());
-        } else {
-            rollupIndexName = request.getRollupIndex();
+    ) {
+        String sourceIndexName = request.getSourceIndex();
+        IndexMetadata sourceIndexMetadata = state.getMetadata().index(sourceIndexName);
+        // Assert source index exists
+        if (sourceIndexMetadata == null) {
+            listener.onFailure(new IndexNotFoundException(sourceIndexName));
+            return;
+        }
+        // Assert source index is a time_series index
+        if (IndexSettings.MODE.get(sourceIndexMetadata.getSettings()) != IndexMode.TIME_SERIES) {
+            listener.onFailure(
+                new ElasticsearchException(
+                    "Rollup requires setting ["
+                        + IndexSettings.MODE.getKey()
+                        + "="
+                        + IndexMode.TIME_SERIES
+                        + "] for index ["
+                        + sourceIndexName
+                        + "]"
+                )
+            );
+            return;
+        }
+        // Assert source index is read-only
+        if (state.blocks().indexBlocked(ClusterBlockLevel.WRITE, sourceIndexName) == false) {
+            listener.onFailure(
+                new ElasticsearchException(
+                    "Rollup requires setting [" + IndexMetadata.SETTING_BLOCKS_WRITE + " = true] for index [" + sourceIndexName + "]"
+                )
+            );
         }
 
-        String tmpIndexName = ".rolluptmp-" + rollupIndexName;
-
-        final XContentBuilder mapping;
-        try {
-            mapping = getMapping(request.getRollupConfig());
-        } catch (IOException e) {
-            listener.onFailure(e);
+        final String rollupIndexName = request.getRollupIndex();
+        if (state.getMetadata().index(rollupIndexName) != null) {
+            listener.onFailure(new ResourceAlreadyExistsException("Rollup index [{}] already exists.", rollupIndexName));
+            return;
+        }
+        final String tmpIndexName = createTmpIndexName(rollupIndexName);
+        if (state.getMetadata().index(tmpIndexName) != null) {
+            listener.onFailure(new ResourceAlreadyExistsException("Temporary rollup index [{}] already exists.", tmpIndexName));
             return;
         }
```

Comment:
Without knowing all the details here, if this action is doing many things that are not atomic then there is room for failure here that needs to be taken into account. But maybe this action should only do the actual rollup and an the rollup ilm action should be changed to do the action listed below here? In case of failure, ILM can retry from the step that has failed.

Reply:
Should we remove the temporary index?

Reply:
Removed temp index in [286cdbd](https://github.com/elastic/elasticsearch/pull/85708/commits/286cdbd3af0dcd6c504285d2f35be661a175ecd0)